File: src/tables/svg.rs
use crate::parser::*;
use crate::Result;

pub fn parse(parser: &mut Parser) -> Result<()> {
    let start = parser.offset();

    parser.read::<u16>("Version")?;
    let list_offset = parser.read::<Offset32>("Offset to the SVG Document List")?.to_usize();
    parser.read::<u32>("Reserved")?;

    parser.jump_to(start + list_offset)?;
    parser.begin_group("SVG Document List");
    let count = parser.read::<u16>("Number of records")?;
    let mut ranges = Vec::new();
    for i in 0..count {
        parser.begin_group(format!("Record {}", i));
        parser.read::<u16>("First glyph ID")?;
        parser.read::<u16>("Last glyph ID")?;
        let offset = parser.read::<Offset32>("Offset to an SVG Document")?.to_usize();
        let size = parser.read::<u32>("SVG Document length")? as usize;
        parser.end_group();

        let doc_start = start + list_offset + offset;
        let doc_end = doc_start + size;
        ranges.push(doc_start..doc_end);
    }
    parser.end_group();

    ranges.sort_by_key(|r| r.start);
    ranges.dedup_by_key(|r| r.start);

    for range in ranges {
        parser.jump_to(range.start)?;

        if parser.peek::<u16>()? == 8075 {
            // Read gzip compressed SVG as is.
            // TODO: decompress
            parser.read_bytes(range.len(), "SVGZ")?;
        } else {
            // Otherwise read as string.
            // According to the spec, it must be in UTF-8, so we are fine.
            parser.read_string(range.len(), "SVG", None)?;
        }
    }

    Ok(())
}

File: src/tables/svg.cpp
#include "src/algo.h"
#include "src/parser.h"
#include "tables.h"

void parseSvg(Parser &parser)
{
    const auto start = parser.offset();

    parser.read<UInt16>("Version");
    const auto listOffset = parser.read<Offset32>("Offset to the SVG Document List");
    parser.read<UInt32>("Reserved");

    parser.advanceTo(start + listOffset);
    parser.beginGroup("SVG Document List");
    const auto count = parser.read<UInt16>("Number of records");
    QVector<Range> ranges;
    for (int i = 0; i < count; ++i) {
        parser.beginGroup(QString("Record %1").arg(i));
        parser.read<UInt16>("First glyph ID");
        parser.read<UInt16>("Last glyph ID");
        const auto offset = parser.read<Offset32>("Offset to an SVG Document");
        const auto size = parser.read<UInt32>("SVG Document length");
        parser.endGroup();

        const auto docStart = start + listOffset + offset;
        const auto docEnd = docStart + size;
        ranges.append({ docStart, docEnd });
    }
    parser.endGroup();

    algo::sort_all_by_key(ranges, &Range::start);
    algo::dedup_vector_by_key(ranges, &Range::start);

    for (const auto &range : ranges) {
        parser.advanceTo(range.start);

        if (parser.peek<UInt16>() == 8075) {
            // Read gzip compressed SVG as is.
            parser.readBytes("SVGZ", range.size());
        } else {
            // Otherwise read as string.
            // According to the spec, it must be in UTF-8, so we are fine.
            parser.readUtf8String("SVG", range.size());
        }
    }
}
